// server/database.js
const fs = require('fs')
const path = require('path')
const Sequelize = require('sequelize')

function associateSchema(db) {
  db.models.events.belongsTo(db.models.contacts, { as: 'contact', foreignKey: 'contact_id' })
  db.models.events.belongsTo(db.models.contacts, { as: 'bookingcontact', foreignKey: 'bookingcontact_id' })
  db.models.events.belongsTo(db.models.venues, { as: 'venue', foreignKey: 'venue_id' })
  db.models.events.belongsTo(db.models.images, { as: 'preferred_image', foreignKey: 'preferred_image_id' })
  db.models.venues.belongsTo(db.models.contacts, { foreignKey: 'venuecontact_id', as: 'venuecontact' })
  db.models.venues.belongsTo(db.models.contacts, { foreignKey: 'addresscontact_id', as: 'addresscontact' })
  db.models.venues.belongsTo(db.models.tags, { foreignKey: 'disabled_id', as: 'disabled' })
  db.models.venues.belongsTo(db.models.tags, { foreignKey: 'toilet_id', as: 'toilet' })
  db.models.venues.belongsTo(db.models.tags, { foreignKey: 'dog_id', as: 'dog' })
  db.models.venues.belongsToMany(db.models.tags, {
    through: 'venue_relatesto_tags',
    foreignKey: 'venue_id',
    otherKey: 'tag_id',
    as: 'regions',
  })
  db.models.tags.belongsToMany(db.models.venues, {
    through: 'venue_relatesto_tags',
    foreignKey: 'tag_id',
    otherKey: 'venue_id',
    as: 'region_venues',
  })
  db.models.tags.belongsTo(db.models.tag_categories, {
    foreignKey: 'tag_category_id',
  })
  db.models.events.belongsToMany(db.models.tags, {
    through: 'event_relatesto_tags',
    foreignKey: 'event_id',
    otherKey: 'tag_id',
    as: 'disciplines',
  })
  db.models.events.belongsToMany(db.models.tags, {
    through: 'event_relatesto_tags',
    foreignKey: 'event_id',
    otherKey: 'tag_id',
    as: 'eventstatus',
  })
  db.models.events.belongsToMany(db.models.tags, {
    through: 'event_relatesto_tags',
    foreignKey: 'event_id',
    otherKey: 'tag_id',
    as: 'eventtypes',
  })
  db.models.events.hasMany(db.models.opening_times, {
    foreignKey: 'event_id',
  })
  return db.sync({ alter: false })
}

function addScopes(db) {
  db.models.events.addScope('defaultScope', {
    attributes: {
      exclude: ['contact_id', 'bookingcontact_id', 'venue_id', 'preferred_image_id'],
    },
    include: [
      'contact',
      'bookingcontact',
      {
        association: 'venue',
        attributes: {
          exclude: ['venuecontact_id', 'addresscontact_id', 'disabled_id', 'toilet_id', 'dog_id'],
        },
        include: [
          'venuecontact',
          'addresscontact',
          {
            association: 'disabled',
            attributes: [ 'id', 'description' ],
          },
          {
            association: 'toilet',
            attributes: [ 'id', 'description' ],
          },
          {
            association: 'dog',
            attributes: [ 'id', 'description' ],
          },
          {
            association: 'regions',
            attributes: [ 'id', 'description' ],
            through: {
              attributes: [],
            },
            include: [
              {
                model: db.models.tag_categories,
                where: { category: 'region' },
                attributes: [],
              },
            ],
          }
        ]
      },
      'preferred_image',
      {
        association: 'disciplines',
        attributes: [ 'id', 'description' ],
        through: {
          attributes: [],
        },
        include: [
          {
            model: db.models.tag_categories,
            where: { category: 'discipline' },
            attributes: [],
          }
        ]
      },
      {
        association: 'eventtypes',
        attributes: [ 'id', 'description' ],
        through: {
          attributes: [],
        },
        include: [
          {
            model: db.models.tag_categories,
            where: { category: 'eventtype' },
            attributes: [],
          }
        ]
      },
      {
        model: db.models.opening_times,
        order: [ 'start' ],
      },
      {
        association: 'eventstatus',
        through: {
          attributes: [],
        },
        include: [
          {
            model: db.models.tag_categories,
            where: { category: 'eventstatus' },
            attributes: [],
          }
        ],
      },
    ],
    order: [ 'title' ]
  }, {
    override: true
  })

  db.models.events.addScope('eventlist', {
    attributes: [ 'id', 'title', 'subtitle', 'shortdesc' ],
    include: [
      'preferred_image',
      {
        association: 'venue',
        include: [ 'regions' ],
      },
      {
        association: 'disciplines',
        attributes: [ 'id', 'description' ],
        through: {
          attributes: [],
        },
        include: [
          {
            model: db.models.tag_categories,
            where: { category: 'discipline' },
            attributes: [],
          },
        ],
      },
      {
        model: db.models.opening_times,
      },
    ],
    order: [ 'title' ],
  })

  db.models.tags.addScope('regions', {
    attributes: ['id', 'description'],
    include: [
      {
        model: db.models.tag_categories,
        attributes: [],
        where: {
          category: 'region',
        }
      }
    ],
    order: [ 'description' ]
  })

  db.models.tags.addScope('disciplines', {
    attributes: ['id', 'description'],
    include: [
      {
        model: db.models.tag_categories,
        attributes: [],
        where: {
          category: 'discipline',
        }
      }
    ],
    order: [ 'description' ]
  })

  db.models.dates.addScope('defaultScope', {
    order: [ 'date' ]
  }, {
    override: true,
  })
}

function stripExtension(file) {
  const matchExtension = /\.[^\\.]+$/
  return file.replace(matchExtension, '')
}

function indexModels(dirname) {
  return new Promise((resolve, reject) => {
    fs.readdir(dirname, (err, files) => {
      if (err) return reject(err);
      resolve(files
        .map(file => path.join(dirname, file))
        .map(stripExtension)
      )
    })
  })
}

async function importSchema(db, models) {
  models.forEach(model => db.import(model))
}

export default async function database(context) {
  const db = new Sequelize(context.dbschema, context.dbuser, context.dbpass, {
    host: context.dbhost,
    dialect: 'mysql',
    operatorsAliases: false,
    logging: false,
  })

  // NB: __dirname is relative to 'dist' directory, due to babel
  return indexModels(path.resolve(__dirname, '../src/node_modules/models'))
    .then(importSchema.bind(null, db))
    .then(associateSchema.bind(null, db))
    .then(addScopes.bind(null, db))
    .then(() => db)
}
