// server/database/index.js
import fs from 'fs'
import path from 'path'
import Sequelize from 'sequelize'
import associations from './associations'
import scopes from './scopes'

function stripExtension(file) {
  const matchExtension = /\.[^\\.]+$/
  return file.replace(matchExtension, '')
}

function indexModels(dirname) {
  return new Promise((resolve, reject) => {
    fs.readdir(dirname, (err, files) => {
      if (err) return reject(err);
      resolve(files
        .map(file => path.join(dirname, file))
        .map(stripExtension)
      )
    })
  })
}

async function importSchema(db, models) {
  models.forEach(model => db.import(model))
}

export default async function database(context) {
  const db = new Sequelize(context.dbschema, context.dbuser, context.dbpass, {
    host: context.dbhost,
    dialect: 'mysql',
    operatorsAliases: false,
    logging: false,
  })

  // NB: __dirname is relative to 'dist' directory, due to babel
  return indexModels(path.resolve(__dirname, '../src/node_modules/models'))
    .then(importSchema.bind(null, db))
    .then(associations.bind(null, db))
    .then(() => db.sync({ alter: false }))
    .then(scopes.bind(null, db))
    .then(() => db)
}
